### PR TITLE
Bubble factorization minigame

### LINKED ISSUE #27: Bubble Game: Create a congratulations screen at the end
- [ ] Task 1: Create a final screen that has a back button and congratulates the user for factorizing all numbers
- [ ] Test 1: UI test that starts game, clicks wrong answer 3 times, then clicks back. Passes if it gets back to home screen

### LINKED ISSUE #28: Bubble Game: Create "Game Over" Page
- [ ] Task 1: create ‘game over’ state when user runs out of hearts.

- [ ] Test 1: Unit test that checks that game_over flag is set when num_hearts == 0

- [ ] Task 2: add 'back' button which redirects user to start menu

- [ ] Test 2: E2E test that checks that user reaches game over screen after completing a run, and that it contains a back button which redirects to start screen

### LINKED ISSUE #43: Bubble Game sub-task: generate multiple rounds per stage
A stage should test the user on multiple numbers.

### PR DESCRIPTION
**Pull request by Waterblokey**
# **Ryan Livengood**
Closes #43 
## Summary:
Added logic for rounds, allowing user to factorize multiple numbers before needing to play again. 
Files to check:
* app/bubble/lib/bubble-context.tsx: Logic for continuing game when not on final round, or ending game once final round is completed.
* app/bubble/lib/bubble-context.test.tsx: Test for player reaching "won" state after factorizing all numbers completely

# **Marlyn Arque Rupa**
Closes #28
Closes #27
## Summary: 
- Created /bubble/end/page.tsx that reads game state (status, lives) from BubbleGameContext to display either "You Win!" or "Game Over" with appropriate styling and messages as part of issue #28 #27.
- Created /bubble/layout.tsx to lift BubbleGameProvider to a shared layout, so game state persists across /bubble/menu, /bubble/game, and /bubble/end.
- Converted all inline CSS styles to Tailwind CSS across all bubble game components (bubble-buttons, bubble-circle, bubble-hearts, game/page, menu/page).
- Integrated TargetCursor animation component from reactbits.dev on menu and game pages.

## Testing:
- Run tests via npm run test 
- Testing includes:
  - End screen shows "Game Over" with status: "lost" after losing all lives
  - End screen shows lives as 0 after game over.
  - End screen shows "You Win!" with status: "won" after factoring the number
  - Play again after game over/winning resets state to "playing" with full lives.
  - Play again generates a new factor and fresh bubbles (board is not empty)
  - Full round-trip cycle: play → lose → play again → play → win

### CODE REVIEW

File: app/bubble/lib/bubble-context.test.tsx [hunk 1]
```diff
@@ -206,3 +207,20 @@ test("resetGame restores lives back to 3", async () => {
   expect(screen.getByTestId("lives")).toHaveTextContent("3");
   expect(screen.getByTestId("status")).toHaveTextContent("playing");
 });
+
+test("player reaches won state after factorizing all numbers", async () => {
```
Comment (ericbinnendyk) [inline app/bubble/lib/bubble-context.test.tsx]: This looks good.

File: app/bubble/lib/bubble-context.tsx [hunk 2]
```diff
@@ -39,6 +41,7 @@ function generateNewGame() {
     factor,
     bubbles: generateBubbles(factor),
     lives: STARTING_LIVES,
+    round: 1,
```
Comment (ericbinnendyk) [inline app/bubble/lib/bubble-context.tsx]: Make sure the rest of the code knows the rounds are 1-indexed instead of 0-indexed. You may want to put this in a comment.

File: app/bubble/lib/bubble-context.tsx [hunk 3]
```diff
@@ -60,9 +63,17 @@ export function BubbleGameProvider({ children }: { children: ReactNode }) {
 
         // Player wins when factor is fully reduced to 1
         if (nextFactor <= 1) {
-          return { ...prev, factor: nextFactor, bubbles: [], status: "won" };
+          if (prev.round >= NUM_ROUNDS) {
+            return { ...prev, factor: nextFactor, bubbles: [], status: "won" };
+          }
+          else {
+            const factor = generateComposite();
```
Comment (ericbinnendyk) [inline app/bubble/lib/bubble-context.tsx]: I think "dividend" or "composite" would be a better word

**Comment by jaewoongy** [comment]
Code Review Assessment:

**Design:**
The code meets our team requirements of developing a second mini-game, called bubble factorization, where it now introduces stages with multiple different randomly generated numbers. 

Regarding common principles, 

**Functionality:**


**Complexity:** 


**Conventions:**


**Testing:**


**Documentation:**

File: app/bubble/components/TargetCursor.tsx [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        style={{ willChange: 'transform' }}
+      />
+      <div
+        className="target-cursor-corner absolute top-1/2 left-1/2 w-3 h-3 border-[3px] border-white -translate-x-[150%] translate-y-1/2 border-r-0 border-t-0"
+        style={{ willChange: 'transform' }}
+      />
+    </div>
+  );
+};
+
```
Comment (jaewoongy) [inline app/bubble/components/TargetCursor.tsx]: Might violate the DRY (Don't repeat yourself) policy here. You can instead define an array and map over them to render divs dynamically.

**Review by jaewoongy — COMMENTED** [review]
Great work! Functionality is great, and tests run successfully on my end. 

The only thing needing addressing are non-functionality related, like code clarity/maintainability principles, but the code runs fine and the design on the site looks great.

Make sure before Merging pull request, we should merge main into the branch to prevent potential file conflicts.


### Summary: 

Code Review Assessment:

**Design:**
The code meets our team requirements of developing a second mini-game, called bubble factorization, where it now introduces stages with multiple different randomly generated numbers. 

Regarding common principles, the developers did a fantastic job with addressing DRY by using components from the components file and re-using their functionality for each page of their mini-game. 

Also, a great design pattern of decoupling the game logic with the UI, making it easier to test, maintain, and add new features later. 

However, the TargetCursor file is quite large and integrates a lot of things at once under one use effect, like mouse movement, event listeners, creating animations, rendering elements. To use the SOLID principle, this can be decoupled into several helper functions, like animation as one helper function, eventlisteners as another, and so forth. This improves not just code clarity, but it will be much easier to maintain and code review, as each helper function will be assigned to a separate purpose. 

**Functionality:** Code looks generally correct, and the whole minigame runs successfully on my end with test cases passing. It also considers edge cases, like making sure the game finishes before lives end or the number goes equal or below 1. 


**Complexity:** I also think it's great that it now transitioned from inlined CSS styes to tailwind CSS makes the code look much cleaner and much easier to understand. 


**Conventions:** Conventions look great using PascalCase and variables seem to be appropriately named. There is some inconsistency with pascal case vs underlined variables but don't think that is an issue at all.


**Testing:**  Testing has been widely utilized, and test cases look good enough to cover a variety of functionalities. However, the TargetCursor has not been tested at all, which may introduce bugs over time given its size.


**Documentation:** 

There are no comments in target-cursor.tsx. Given the size of the file, in order to improve clarity for another user to make edits, or in case something fails, it would be very difficult to maintain this codebase without proper documentation. The end and game page.tsx files could also use some comments, but should be very minimal. Bubble-context.tsx and game-logic.tsx did a fantastic job with commenting all functionalities for each function. 

The tests are well-documented, as each individual unit test describes what it expects to look in the test.
